src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +429,88 @@ data CodingFailureMode
 replacementChar :: Char
 replacementChar = '\xFFFD'
 
+data UTF8CharDecodeState a
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
+
+{-# INLINE parseCharUtf8WithD' #-}
+parseCharUtf8WithD' ::
+       Monad m => CodingFailureMode -> ParserD.Parser m Word8 Char
+parseCharUtf8WithD' cfm =
+    let A.Array _ ptr _ = utf8d
+    in ParserD.Parser (step' ptr) initial extract
+
+    where
+
+    prefix = "Streamly.Internal.Data.Stream.parseCharUtf8WithD:"
+
+    {-# INLINE initial #-}
+    initial = return $ ParserD.IPartial UTF8CharDecodeInit
+
+    handleError table e s =
+        case cfm of
+            ErrorOnCodingFailure -> return $ ParserD.Error e
+            TransliterateCodingFailure ->
+                return $ case s of
+                    Just _ -> ParserD.Done 1 replacementChar
+                    Nothing -> ParserD.Continue 0 UTF8CharDecodeInit
+            DropOnCodingFailure ->
+                case s of
+                    Just x -> step' table UTF8CharDecodeInit x
+                    Nothing -> return $ ParserD.Partial 0 UTF8CharDecodeInit
```

Comment:
This should be `Continue 0 UTF8CharDecodeInit` as there is no usable partial result being produced.

Reply:
after removing `step'` the handleError function should no longer be monadic.